tun2socks/lwip/device.go

```diff
@@ -0,0 +1,145 @@
+	if sd == nil || pl == nil {
+		return nil, errors.New("both sd and pl are required")
+	}
+
+	t2s := &LwIPTun2SocksDevice{
+		tcp:     newTCPHandler(sd),
+		udp:     newUDPHandler(pl, 30*time.Second),
+		stack:   lwipLib.NewLWIPStack(),
+		outPipe: newPacketPipe(packetMTU),
+	}
+	lwipLib.RegisterTCPConnHandler(t2s.tcp)
+	lwipLib.RegisterUDPConnHandler(t2s.udp)
+	lwipLib.RegisterOutputFn(t2s.handleIPResponseFromStack)
+
+	return t2s, nil
+}
+
+// Close implements io.Closer and network.IPDevice. It closes the device, rendering it unusable for I/O.
+//
+// Close does not close other objects that are passed to the device, such as StreamDialer, PacketListener, Writer, or
+// Reader. You are responsible for closing these objects yourself.
+func (t2s *LwIPTun2SocksDevice) Close() error {
+	err := t2s.stack.Close()
+	t2s.outPipe.close()
+	if err != nil {
+		return fmt.Errorf("failed to close lwIP stack: %w", err)
+	}
+	return nil
+}
+
+// MTU implements network.IPDevice. It returns the maximum buffer size of a single IP packet that can be processed by
+// this device.
+func (t2s *LwIPTun2SocksDevice) MTU() int {
+	return packetMTU
+}
+
+// lwIP responses from TCP/UDP connections (in IP Packets) callback function
+func (t2s *LwIPTun2SocksDevice) handleIPResponseFromStack(b []byte) (int, error) {
+	return t2s.outPipe.write(b)
+}
+
+// Read implements io.Reader and network.IPDevice. It reads one IP packet from the TCP/UDP response, blocking until a
+// packet arrives or this device is closed. To prevent potential memory allocations, use the WriteTo function.
+func (t2s *LwIPTun2SocksDevice) Read(p []byte) (int, error) {
+	return t2s.outPipe.read(p)
+}
+
+// WriteTo implements io.WriterTo. It writes all IP packets from TCP/UDP responses to `w` until all data is written or
+// an error occurs. This function will not allocate any intermediate buffers.
+//
+// WriteTo returns the total number of bytes written and any error encountered during the write.
+func (t2s *LwIPTun2SocksDevice) WriteTo(w io.Writer) (int64, error) {
+	return t2s.outPipe.writeTo(w)
+}
+
+// Write implements io.Writer and network.IPDevice. It writes a single IP packet to this device. The device will then
+// translate the IP packet into a TCP or UDP traffic.
+func (t2s *LwIPTun2SocksDevice) Write(b []byte) (int, error) {
+	n, err := t2s.stack.Write(b)
+	// workaround: lwip netstack did not use a standard error code
+	if err != nil && err.Error() == "stack closed" {
+		return n, net.ErrClosed
+	}
+	return n, err
+}
+
+// ReadFrom implements io.ReaderFrom. It reads all IP packets from `r` until EOF or error. EOF can occur if either this
+// device is closed or `r` is closed, and it will not be treated as an error. This function will only allocate a single
+// buffer of MTU() bytes.
+//
+// ReadFrom returns the total number of bytes read and any error encountered during the read.
+func (t2s *LwIPTun2SocksDevice) ReadFrom(r io.Reader) (int64, error) {
+	p := make([]byte, t2s.MTU())
```

Comment:
BTW, allocating a single buffer for all the VPN is perfectly fine. It's more of an issue when there are large buffers per connection or UDP packet. So perhaps we don't need to worry too much about the ReadFrom optimization.

Reply:
For the buffer thing, yes you are right. But here I'm trying to add EOF handling here. Because if you use `io.Copy` or `io.CopyBuffer`, and when we close the `tun` device / `tun2socks` device, it  will return an error. I prefer to return EOF instead of an error in these scenario, because these `close()` operations are expected instead of errors.

Reply:
In that case, you can still call io.CopyBuffer and handle the closed error in the caller. Or make Read return EOF after close?

Reply:
We already did that. It's not our `Read` returns error, it's the `tun` device (as I mentioned in the above comment) implementation that doesn't follow the go standard and returns an `os.ErrClosed`. But as I said above, we can always add a wrapper around it.